src/Agent.Plugins/BuildDrop/BuildDropPlugin.cs

```diff
@@ -0,0 +1,191 @@
+    // Can be invoked from a build run or a release run should a build be set as the artifact. 
+    public class DownloadBuildDropTask : BuildDropTaskPluginBase
+    {
+        // Same as https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/DownloadBuildDropV0/task.json
+        public override Guid Id => new Guid("61F2A582-95AE-4948-B34D-A1B3C4F6A737");
+
+        protected override async Task ProcessCommandInternalAsync(
+            AgentTaskPluginExecutionContext context, 
+            string targetPath, 
+            Guid projectId,
+            string artifactName,
+            CancellationToken token)
+        {
+            // Create target directory if absent
+            string fullPath = Path.GetFullPath(targetPath);
+            bool isDir = Directory.Exists(fullPath);
+            if (!isDir)
+            {
+                Directory.CreateDirectory(fullPath);
+            }
+
+            // Build ID
+            int buildId = 0;
+            string buildIdStr = string.Empty;
+            // Determine the build id
+            if (context.Inputs.TryGetValue(ArtifactEventProperties.BuildId, out buildIdStr) && 
+                Int32.TryParse(buildIdStr, out buildId) &&
+                buildId != 0)
+            {
+                // A) Build Id provided by user input
+                context.Output($"Download from the specified build: #{ buildId }");
+            }
+            else
+            {
+                // B) Build Id provided by environment
+                buildIdStr = context.Variables.GetValueOrDefault(BuildVariables.BuildId)?.Value ?? string.Empty;
+                if (int.TryParse(buildIdStr, out buildId) &&
+                    buildId != 0)
+                {
+                    context.Output($"Download from the current build: #{ buildId }");
+                }
+                else
+                {
+                    string hostType = context.Variables.GetValueOrDefault("system.hosttype")?.Value; 
+                    if (string.Equals(hostType, "Release", StringComparison.OrdinalIgnoreCase)) {
+                        throw new InvalidOperationException(
+                            $"Trying to download build drop in '{ hostType ?? string.Empty }' environment but build id is not present. " + 
+                            $"Can only download a build drop in '{ hostType ?? string.Empty }' environment if the artifact is a build."); 
+                    } else if (!string.Equals(hostType, "Build", StringComparison.OrdinalIgnoreCase)) {
+                        throw new InvalidOperationException(
+                            $"Cannot download a build drop from '{ hostType ?? string.Empty }' environment."); 
+                    } else {
+                        // This should not happen since the build id comes from build environment. But a user may override that so we must be careful.
+                        throw new ArgumentException($"Build Id is not valid: #{ buildIdStr }");
```

Comment:
LOC all exception strings.

as long as the string will shows with System.debug =false, you need to loc them